### PR TITLE
Service components matchers

### PR DESCRIPTION
**Pull request by mapingo**

### CODE REVIEW

File: example-context/example-service/example-command/example-command-handler/src/test/java/uk/gov/justice/services/example/cakeshop/command/handler/MakeCakeCommandHandlerTest.java [hunk 1]
```diff
@@ -34,6 +38,12 @@
     private MakeCakeCommandHandler makeCakeCommandHandler;
 
     @Test
+    public void shouldHaveCorrectHandlesAnnotation() throws Exception {
+        assertThat(makeCakeCommandHandler, isHandler(COMMAND_HANDLER)
+                .with(method("handle").thatHandles("example.make-cake")));
```
Comment (matt-rich) [inline example-context/example-service/example-command/example-command-handler/src/test/java/uk/gov/justice/services/example/cakeshop/command/handler/MakeCakeCommandHandlerTest.java]: is the method really called "handle" ? perhaps we should refactor that...

File: example-context/example-service/example-query/example-query-controller/src/test/java/uk/gov/justice/services/example/cakeshop/query/controller/RecipesQueryControllerTest.java [hunk 2]
```diff
@@ -1,14 +1,26 @@
...
+import static uk.gov.justice.services.core.annotation.Component.QUERY_CONTROLLER;
+import static uk.gov.justice.services.test.utils.core.matchers.HandlerClassMatcher.isHandlerClass;
+import static uk.gov.justice.services.test.utils.core.matchers.HandlerMethodMatcher.method;
 
 import org.junit.Test;
 
 public class RecipesQueryControllerTest {
 
     @Test
     public void shouldHandleRecipesQuery() throws Exception {
```
Comment (matt-rich) [inline example-context/example-service/example-query/example-query-controller/src/test/java/uk/gov/justice/services/example/cakeshop/query/controller/RecipesQueryControllerTest.java]: as above

File: test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/matchers/HandlerClassMatcherTest.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    @Test
+    public void shouldMatchHandlerMethod() throws Exception {
+        assertThat(ValidCommandApi.class, isHandlerClass(COMMAND_API).with(method("testA").thatHandles("testA")));
+    }
+
+    @Test(expected = AssertionError.class)
+    public void shouldNotMatchWhenNoHandlerMethod() throws Exception {
+        assertThat(NoHandlerMethod.class, isHandlerClass(COMMAND_API).with(method("testA").thatHandles("testA")));
```
Comment (matt-rich) [inline test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/matchers/HandlerClassMatcherTest.java]: maybe update the handles string to context.commandA or something to make it clear that it is a handlesAnnotation type thing. for these 2 tests

File: test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/matchers/HandlerMatcherTest.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    @Test(expected = AssertionError.class)
+    public void shouldNotMatchInstanceOfAHandlerWithTheWrongAnnotation() throws Exception {
+        assertThat(new ValidCommandApi(), isHandler(QUERY_API));
+    }
+
+    @Test
+    public void shouldMatchWithGivenMatcherHandlerMethod() throws Exception {
+        assertThat(new ValidCommandApi(), isHandler(COMMAND_API).with(method("testA").thatHandles("testA")));
```
Comment (matt-rich) [inline test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/matchers/HandlerMatcherTest.java]: as above, but just a suggestion...

File: test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/matchers/MethodHandlesAnnotationMatcherTest.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+import uk.gov.justice.services.core.annotation.Handles;
+
+import org.junit.Test;
+
+public class MethodHandlesAnnotationMatcherTest {
+
+    @Test
+    public void shouldMatchMethodWithHandlesAnnotation() throws Exception {
+        assertThat(TestClass.class.getMethod("testA"), methodThatHandles("testA"));
```
Comment (matt-rich) [inline test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/matchers/MethodHandlesAnnotationMatcherTest.java]: again, as above.

File: test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/matchers/HandlerClassMatcher.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+        if (matcher.isPresent() && !matcher.get().matches(handlerClass)) {
+            matcher.get().describeMismatch(handlerClass, description);
+            return false;
+        }
+
+        return true;
+    }
+
+    private boolean isNotServiceComponent(final Class<?> handlerClass) {
```
Comment (mapingo) [inline test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/matchers/HandlerClassMatcher.java]: I could not find anything...but I have refactored this code

**Pull request merged**